File: backend/src/main/kotlin/challenge/Payment.kt
package challenge

import java.util.Date


interface PaymentMethod

data class CreditCard(val number: String) : PaymentMethod

data class Payment(val order: Order, val paymentMethod: PaymentMethod) {
    val paidAt = Date()
    val authorizationNumber = paidAt.time
    val amount = order.totalAmount

    val invoice = Invoice(order)
}

File: backend/individual-assignment/typescript/src/payment.ts
import { Invoice } from "./invoice";
import { Order } from "./order";

export interface PaymentMethod {}

export class Payment {
  public paidAt: Date;
  public authorizationNumber: number;
  public amount: number;
  public invoice: Invoice;

  constructor(
    public order: Order,
    public paymentMethod: PaymentMethod,
  ) {
    this.paidAt = new Date();
    this.authorizationNumber = this.paidAt.getTime();
    this.amount = order.totalAmount;
    this.invoice = new Invoice(order);
  }
}

export class CreditCard implements PaymentMethod {
  constructor(public number: string) {}
}
